### PR TITLE
Added a unit test for AdminServlet

### PR DESCRIPTION
**Pull request by clague17**
Testing the entirety of AdminServlet.

Tested the statistics are refreshed appropriately by generating mock Users, Messages, and Conversations.
Tested the administrators can add new administrators.

### CODE REVIEW

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  @Test
+  public void testDoGet() throws IOException, ServletException {
+    adminServlet.doGet(mockRequest, mockResponse);
+    mockRequest.getParameter("username");
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+  @Test
+  public void testGetStats() throws IOException, ServletException {
+    Mockito.when(mockRequest.getParameter("username")).thenReturn(null);
```
Comment (Gonzo717) [inline src/test/java/codeu/controller/AdminServletTest.java]: Explain this line, does it mean that when you request the parameter "username", you will get null in return? If yes, why is this needed?

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    Mockito.when(mockRequest.getParameter("username")).thenReturn("test username");
+
+    adminServlet.doPost(mockRequest, mockResponse);
+
+    Mockito.verify(mockRequest).setAttribute("success", "Admin successfully added");
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+  // @Test
+  // public void testDoPost_ExistingUser() throws IOException, ServletException {
```
Comment (Gonzo717) [inline src/test/java/codeu/controller/AdminServletTest.java]: Why did you comment this out?

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    Mockito.when(mockRequest.getParameter("username")).thenReturn(null);
+
+    UUID fakeConversationId = UUID.randomUUID();
+    List<Conversation> fakeConversationList = new ArrayList<>();
+    Conversation fakeConversation =
+        new Conversation(fakeConversationId, UUID.randomUUID(), "test_conversation", Instant.now());
+    
+    fakeConversationList.add(fakeConversation);
+
+    Mockito.when(mockConvoStore.getAllConversations())
```
Comment (cmondragon15117) [inline src/test/java/codeu/controller/AdminServletTest.java]: Like Gonzalo, I'm still trying to understand Mockito so most of my comments will also be questions. So what are you trying to accomplish with line 96? Is it just saying that when it gets all the conversations it returns the fake conversations?

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    
+    Mockito.when(mockRequest.getSession()).thenReturn(mockSession);
+
+    adminServlet.doPost(mockRequest, mockResponse);
+
+    Mockito.verify(mockSession).setAttribute("numUsers", numTotalUsers);
+    Mockito.verify(mockSession).setAttribute("numMessages", numTotalMessages);
+    Mockito.verify(mockSession).setAttribute("numConvos", numTotalConvos);
+
+    Mockito.verify(mockResponse).sendRedirect("/admin");
```
Comment (cmondragon15117) [inline src/test/java/codeu/controller/AdminServletTest.java]: Can you clarify why this redirects to the admin page?

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+    mockUserStore = Mockito.mock(UserStore.class);
+    mockConvoStore = Mockito.mock(ConversationStore.class);
+    mockMessageStore = Mockito.mock(MessageStore.class);
+
+    User fakeUser =
+        new User(
+            UUID.randomUUID(),
+            "test username",
+            "$2a$10$bBiLUAVmUFK6Iwg5rmpBUOIBW6rIMhU1eKfi3KR60V9UXaYTwPfHy",
```
[the remark this answers is not in the document]
Comment (cmondragon15117) [reply src/test/java/codeu/controller/AdminServletTest.java]: I was wondering the same thing

File: src/test/java/codeu/controller/AdminServletTest.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            Instant.now());
```
Comment (cmondragon15117) [inline src/test/java/codeu/controller/AdminServletTest.java]: Just a question, what is the Instant.now() used for?

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (clague17) [reply src/test/java/codeu/controller/AdminServletTest.java]: So I looked at the Mockito test for stuff like the LoginServlet, and I literally copy pasted this fake user instantiation along with the hashed password

[the remark this answers is not in the document]
Comment (Gonzo717) [reply src/test/java/codeu/controller/AdminServletTest.java]: RIP

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 1 above)
Comment (clague17) [reply src/test/java/codeu/controller/AdminServletTest.java]: Yes thats exactly what it does. This is necessary because of the way I formatted the AdminServlet. Essentially, once a button is pushed (form is submitted) on admin.jsp, the doPost method is called. Right now I have two things going on in admin.jsp once doPost is called: Adding a potential admin (IF the username parameter is not null) and/or updating the site stats. If the username parameter is null, then don't add another admin, but if it is, which is what I am testing here, then don't add an admin, just update the site stats.

Does that kinda make sense?

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 2 above)
Comment (clague17) [reply src/test/java/codeu/controller/AdminServletTest.java]: I just commented out this test because I copy pasted the whole thing from LoginServletTest before I edited it. I just forgot to delete this last test, but its irrelevant.

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 3 above)
Comment (clague17) [reply src/test/java/codeu/controller/AdminServletTest.java]: Cool yeah thats exactly what its doing. In order to test this, I had to create a fake conversation store along with a fake conversation with fake messages and fake users. It got really involved real quick.

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 2 above)
Comment (cmondragon15117) [reply src/test/java/codeu/controller/AdminServletTest.java]: I would just delete it so the code looks cleaner

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 6 above)
Comment (Gonzo717) [reply src/test/java/codeu/controller/AdminServletTest.java]: So I had the exact same question, but if you look inside the User class in the path main/java/codeu/model/data, you can see that a User consists of 4 parameters which are: an ID,a name, a password hash, an a parameter named "creation." My guess is that creation stores the time the user object was made, so passing Instant.now() for that parameter gives it the current time from your system clock.

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 4 above)
Comment (clague17) [reply src/test/java/codeu/controller/AdminServletTest.java]: This should redirect to the admin page because all the method does is update the admin stats so there's no real reason to go anywhere else.

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 6 above)
Comment (clague17) [reply src/test/java/codeu/controller/AdminServletTest.java]: Instant.now is just a time stamp of when the user was created. I am using it here I think just because it is necessary as per the constructor of the User but its not really useful outside of that.

Actually yes it is useful. I am going to use that to implement a "recently registered users" type of statistic for the admin page. Great suggestion haha thanks!

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 1 above)
Comment (Gonzo717) [reply src/test/java/codeu/controller/AdminServletTest.java]: Got it, thanks for explaining.

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 2 above)
Comment (clague17) [reply src/test/java/codeu/controller/AdminServletTest.java]: Yeah I have it deleted on the branch im currently working on so once I push that later we'll be all set

File: src/test/java/codeu/controller/AdminServletTest.java (hunk 3 above)
Comment (cmondragon15117) [reply src/test/java/codeu/controller/AdminServletTest.java]: Ah okay, thanks!
